overlord/state/warning.go

```diff
@@ -0,0 +1,280 @@
+	DefaultExpireAfter = time.Hour * 24 * 28
+
+	errNoMessage     = errors.New("warning has no message")
+	errBadMessage    = errors.New("malformed warning message")
+	errNoFirstAdded  = errors.New("warning has no first-added timestamp")
+	errNoExpireAfter = errors.New("warning has no expire-after duration")
+	errNoRepeatAfter = errors.New("warning has no repeat-after duration")
+)
+
+type jsonWarning struct {
+	Message     string     `json:"message"`
+	FirstAdded  time.Time  `json:"first-added"`
+	LastAdded   time.Time  `json:"last-added"`
+	LastShown   *time.Time `json:"last-shown,omitempty"`
+	ExpireAfter string     `json:"expire-after,omitempty"`
+	RepeatAfter string     `json:"repeat-after,omitempty"`
+}
+
+type Warning struct {
+	// the warning text itself. Only one of these in the system at a time.
+	message string
+	// the first time one of these messages was created
+	firstAdded time.Time
+	// the last time one of these was created
+	lastAdded time.Time
+	// the last time one of these was shown to the user
+	lastShown time.Time
+	// how much time since one of these was last added should we drop the message
+	expireAfter time.Duration
+	// how much time since one of these was last shown should we repeat it
+	repeatAfter time.Duration
+}
+
+func (w *Warning) String() string {
+	return w.message
+}
+
+func (w *Warning) MarshalJSON() ([]byte, error) {
+	jw := jsonWarning{
+		Message:     w.message,
+		FirstAdded:  w.firstAdded,
+		LastAdded:   w.lastAdded,
+		ExpireAfter: w.expireAfter.String(),
+		RepeatAfter: w.repeatAfter.String(),
+	}
+	if !w.lastShown.IsZero() {
+		jw.LastShown = &w.lastShown
+	}
+
+	return json.Marshal(jw)
+}
+
+func (w *Warning) UnmarshalJSON(data []byte) error {
+	var jw jsonWarning
+	err := json.Unmarshal(data, &jw)
+	if err != nil {
+		return err
+	}
+	w.message = jw.Message
+	w.firstAdded = jw.FirstAdded
+	w.lastAdded = jw.LastAdded
+	if jw.LastShown != nil {
+		w.lastShown = *jw.LastShown
+	}
+	if jw.ExpireAfter != "" {
+		w.expireAfter, err = time.ParseDuration(jw.ExpireAfter)
+		if err != nil {
+			return err
+		}
+	}
+	if jw.RepeatAfter != "" {
+		w.repeatAfter, err = time.ParseDuration(jw.RepeatAfter)
+		if err != nil {
+			return err
+		}
+	}
+
+	return w.validate()
+}
+
+func (w *Warning) validate() (e error) {
+	if w.message == "" {
+		return errNoMessage
+	}
+	if strings.TrimSpace(w.message) != w.message {
+		return errBadMessage
+	}
+	if w.firstAdded.IsZero() {
+		return errNoFirstAdded
+	}
+	if w.expireAfter == 0 {
+		return errNoExpireAfter
+	}
+	if w.repeatAfter == 0 {
+		return errNoRepeatAfter
+	}
+	return nil
+}
+
+func (w *Warning) Expired(now time.Time) bool {
+	return w.lastAdded.Add(w.expireAfter).Before(now)
+}
+
+func (w *Warning) IsShowable(t time.Time) bool {
+	return (w.lastShown.IsZero() || w.lastShown.Add(w.repeatAfter).Before(t)) && !w.firstAdded.After(t)
+}
+
+// flattenWarning loops over the warnings map, and returns all
+// warnings therein as a flat list, for serialising.
```

Comment:
The choice of a map for the messages feels a bit unusual. The slice feels slightly more natural for the in memory representation, and while it would be more expensive to find strings, it would avoid the cost of flattening it.

With that said, this is not a blocker as it's just the in-memory representation and easy to go one way or the other later as we learn more. So let's go with your choice.